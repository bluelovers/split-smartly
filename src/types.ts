import splitSmartly, { EnumIncludeSeparatorMode, SearchResults } from './index';

export type IBracketsItem = [open: IBracketsObject["open"], close: IBracketsObject["close"], searchLevels?: IBracketsObject["searchLevels"], ignoreMode?: IBracketsObject["ignoreMode"]];

export type IBrackets = IBracketsItem[];

export type IBracketsInput = true | string | IBrackets | Record<string, string>;

export interface IBracketsObject
{
	open: string;
	close: string;
	searchLevels: true | number[];
	ignoreMode: boolean;
	openPosition?: number;
}

export interface IBracketsMap
{
	[k: string]: IBracketsObject
}

export type IMentionsRecord = Record<string, string>;

export type IMentionsInput = string | boolean | string[] | IMentionsRecord;

export interface ISplitSettings<M extends IIncludeSeparatorMode>
{
	brackets: IBracketsInput;
	mentions: IMentionsInput;
	ignoreInsideQuotes: boolean;
	includeSeparatorMode: M;
	ignoreCase: boolean;
	trimResult: boolean;
	trimSeparators: boolean;

	check(checkParams: ICheckParams): boolean;

	defaultBrackets: IBrackets;
	separators: ISeparators;

	init(): ISplitSettings<M>;

	merge<M2 extends IIncludeSeparatorMode = M>(settings: ISplitSettingsInput<M2>): ISplitSettings<M2>;

	arrayToPattern(arr: string[]): string;

	createRegExp(pattern: string): RegExp;

	createBracketsMap(): ISplitSettings<M>;

	createBracketsSearch(): ISplitSettings<M>;

	createSeparatorsSearch(): ISplitSettings<M>;

	bracketsMap: IBracketsMap;
	bracketsSearch: RegExp;
	separatorSearch: RegExp;
	searchWithin: boolean
	indexes: number | number[]
	returnIterator: boolean

	includePositions: boolean,
}

export interface ISplitSettingsInput<M extends IIncludeSeparatorMode> extends Partial<ISplitSettings<M>>
{

}

export interface ISearchSettings<M extends IIncludeSeparatorMode = EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_SEPARATELY> extends ISplitSettings<M>
{

}

export interface ISearchSettingsInput<M extends IIncludeSeparatorMode = EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_SEPARATELY> extends Partial<ISearchSettings<M>>
{

}

export interface ICheckParams
{
	getString(): string;
	getTextAfter(): string;
	getMentions(): string[];
	getSeparator(): string;

	readonly string: string;
	readonly textAfter: string;
	readonly mentions: string[];
	readonly separator: string;
}

export interface IMention
{
	index: number
	mention: string
}

export type IPipeItem = string | [string, ISeparators] | [ISeparators, string] | ISeparators

interface IPipeItemMap
{
	[EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_SEPARATELY]: string
	[EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_LEFT]: [string, ISeparators]
	[EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_RIGHT]: [ISeparators, string]
	[EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_ONLY]: ISeparators
}

export type IGetPipeItemByIncludeSeparatorMode<M extends IIncludeSeparatorMode> =
	M extends EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_ONLY ?
		IPipeItemMap[EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_ONLY] :
		M extends EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_RIGHT ?
			IPipeItemMap[EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_RIGHT] :
			M extends EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_LEFT ?
				IPipeItemMap[EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_LEFT] :
				IPipeItemMap[EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_SEPARATELY]
	;
export type IGetIncludeSeparatorModeBySettings<T extends ISearchSettingsInput<IIncludeSeparatorMode>> = T extends ISearchSettingsInput<infer M>
	? M
	: IIncludeSeparatorMode
export type IGetPipeItemBySettings<T extends ISearchSettingsInput<IIncludeSeparatorMode>> = IGetPipeItemByIncludeSeparatorMode<IGetIncludeSeparatorModeBySettings<T>>
export type IIncludeSeparatorMode =
	EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_NONE
	| EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_SEPARATELY
	| EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_LEFT
	| EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_RIGHT
	| EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_ONLY
export type ISeparators = string | string[] | RegExp;
export type ISeparatorsNode = ISeparators | ITextNodeSeparator;

export interface ITextNodeBase
{
	text: string
	position?: number
	isSeparator?: never
	mentions?: string[]
}

export interface ITextNodeSeparator extends Omit<ITextNodeBase, 'isSeparator' | 'text'>
{
	text: ISeparators
	position: number
	isSeparator: true
}

export type ITextNodeOrText = string | ITextNodeBase | ITextNodeSeparator;

export const enum EnumFindBracketsAction
{
	ACTION_CLOSE = 1,
	ACTION_OPEN,
	ACTION_ADD_FRAGMENT,
	ACTION_NULL,
}

export type IParametersSplitSmartlyReturnQuery<M extends IIncludeSeparatorMode> = [Exclude<ISeparators, string>, ISplitSettingsInput<M>?];

export type IParametersSplitSmartlyReturnResult<M extends IIncludeSeparatorMode> = [...([string, ISeparators?] | [string]), ISplitSettingsInput<M>?];

export type IParametersSplitSmartly<M extends IIncludeSeparatorMode> = IParametersSplitSmartlyReturnQuery<M> | IParametersSplitSmartlyReturnResult<M>;

export interface ISplitFunctionCore<M extends IIncludeSeparatorMode>
{
	(string: string, settings?: ISearchSettingsInput<M>): SearchResults<M> | IGetPipeItemByIncludeSeparatorMode<M> | IGetPipeItemByIncludeSeparatorMode<M>[]
}

export interface ISplitFunction<M extends IIncludeSeparatorMode> extends ISplitFunctionCore<M>, ThisType<ISplitSettings<M>>
{
	getOne<T extends ISearchSettingsInput<IIncludeSeparatorMode> = ISearchSettingsInput<M>>(string: string, index: number, settings?: T): IGetPipeItemBySettings<T>;
	getFirst<T extends ISearchSettingsInput<IIncludeSeparatorMode> = ISearchSettingsInput<M>>(string: string, settings?: T): IGetPipeItemBySettings<T>;
	getIndexes<T extends ISearchSettingsInput<IIncludeSeparatorMode> = ISearchSettingsInput<M>>(string: string, indexes: any[], settings?: T): IGetPipeItemBySettings<T>;
	getIterator<T extends ISearchSettingsInput<IIncludeSeparatorMode> = ISearchSettingsInput<M>>(string: string, settings?: T): SearchResults<IGetIncludeSeparatorModeBySettings<T>>;
}

export interface IParameterSeparator extends RegExpExecArray
{
	searchWithinData?: IBracketsObject
}

export type IReturnTypeCheckSeparator = [text: string | ITextNodeBase, separator: ISeparatorsNode, checked: boolean];
