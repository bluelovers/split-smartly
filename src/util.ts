import { ITextNodeOrText } from './types';

export const once = <T extends (...args: any[]) => any>(fn: T) =>
{
	let value: any, hasValue: any
	return function (...args)
	{
		if (!hasValue)
		{
			value = fn(...args)
			hasValue = true
		}
		return value
	} as T
}
export const isEmpty = <T extends ITextNodeOrText | any[] | Record<any, any>>(value: T) =>
{
	if (!value) return true

	if (Array.isArray(value))
	{
		if (value.length === 0) return true
	}

	else if (typeof value === 'object')
	{
		if (Object.keys(value).length === 0) return true
	}

	return false
}
export const first = <T>(value: T[]): T => value[0]
export const last = <T>(value: T[]): T => value[value.length - 1]
