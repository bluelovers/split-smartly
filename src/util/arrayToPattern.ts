
let screenedSymbols: Set<string>;

export function arrayToPattern(arr: (RegExp | string)[])
{
	screenedSymbols ??= new Set('.{}[]^()+*?\\/$|'.split(''))

	return arr
		.map(s => {

			if (s instanceof RegExp)
			{
				return s.source;
			}

			return s.split('').map(s => screenedSymbols.has(s) ? '\\' + s : s).join('')
		})
		.join('|')
}
