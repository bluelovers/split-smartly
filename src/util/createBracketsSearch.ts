import { IIncludeSeparatorMode, ISplitSettings } from '../types';

export function createBracketsSearch<M extends IIncludeSeparatorMode>(settings: ISplitSettings<M>)
{
	const patternParts = Object.entries(settings.bracketsMap)
		// @ts-ignore
		.flatMap(([, { close, open }]) => close !== open ? [open, close] : open)
		.concat(Object.keys(settings.mentions || {}))
		.filter(Boolean)

	const pattern = settings.arrayToPattern(patternParts)
	settings.bracketsSearch = settings.createRegExp(pattern)

	return settings
}
