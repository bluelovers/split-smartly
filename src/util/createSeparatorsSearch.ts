import { IIncludeSeparatorMode, ISplitSettings } from '../types';

export function createSeparatorsSearch<M extends IIncludeSeparatorMode>(settings: ISplitSettings<M>)
{
	const { separators } = settings

	if (typeof separators === 'string' || Array.isArray(separators))
	{
		const pattern = settings.arrayToPattern([separators].flat().filter(Boolean))
		settings.separatorSearch = settings.createRegExp(pattern)
	}
	else if (separators)
	{
		settings.separatorSearch = separators
		settings.ignoreCase = separators.ignoreCase
	}
	else
	{
		settings.separatorSearch = /empty/
	}

	return settings
}
