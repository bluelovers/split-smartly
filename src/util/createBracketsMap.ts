import { IBrackets, IBracketsMap, IIncludeSeparatorMode, ISplitSettings } from '../types';
import { first } from '../util';

export function createBracketsMap<M extends IIncludeSeparatorMode>(settings: ISplitSettings<M>)
{
	let { brackets = [] as IBrackets, ignoreInsideQuotes } = settings
	if (brackets === true)
	{
		brackets = settings.defaultBrackets
	}
	else if (typeof brackets === 'object' && !Array.isArray(brackets))
	{
		brackets = Object.entries(brackets) as IBrackets
	}
	else if (typeof brackets === 'string')
	{
		brackets = brackets
			.split(',')
			.map(pairText =>
			{
				let pair = pairText.trim().split(' ')
				if (pair.length !== 2)
				{
					if (first(pair).length === 2)
					{
						pair = first(pair).split('')
					}
					else
					{
						throw new TypeError(`open and close parts of brackets should be separated by space symbol`)
					}
				}
				return pair as any
			}) as IBrackets
	}

	if (ignoreInsideQuotes)
	{
		brackets.unshift([`'`, , , true], [`"`, , , true])
	}

	settings.bracketsMap = brackets.reduce((map, [open, close, ...args]) =>
	{
		if (args.length === 1 && !settings.searchWithin)
		{
			args.unshift(undefined)
		}
		let [searchLevels = settings.searchWithin && 1, ignoreMode] = args
		if (typeof searchLevels === 'number')
		{
			searchLevels = [searchLevels]
		}
		map[open] = { open, ignoreMode, searchLevels, close: close || open }
		return map
	}, {} as IBracketsMap)

	return settings
}
