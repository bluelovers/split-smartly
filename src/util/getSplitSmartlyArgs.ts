import {
	IIncludeSeparatorMode,
	IParametersSplitSmartly,
	ISeparators,
	ISplitSettings,
	ISplitSettingsInput,
} from '../types';
import { first } from '../util';

export const getSplitSmartlyArgs = <M extends IIncludeSeparatorMode, M2 extends IIncludeSeparatorMode = M>(args: IParametersSplitSmartly<M>,
	extraSettings?: ISplitSettingsInput<M2>,
): [string, ISeparators, ISplitSettings<M2>] =>
{

	if (!args?.length)
	{
		throw new RangeError('empty arguments')
	}
	else if (args.length === 3)
	{
		if (!extraSettings) return args as any
	}
	else if (args.length === 1)
	{
		const arg = first(args)
		if (typeof arg === 'string')
		{
			// @ts-ignore
			args.push(',', {})
		}
		else if (Array.isArray(arg))
		{
			args.unshift(null)
			args.push({})
		}
		else if (typeof arg === 'object')
		{
			// @ts-ignore
			args.unshift(null, ',')
		}
	}
	else if (args.length === 2)
	{
		if (typeof args[0] === 'string' && args[1] instanceof RegExp)
		{
			args.push({})
		}
		else if (typeof args[1] === 'string' || Array.isArray(args[1]))
		{
			args.push({})
		}
		else
		{
			args.unshift(null)
		}
	}
	else if (args.length > 3)
	{
		throw new RangeError('Too much arguments passed to splitSmartly function!!!')
	}

	// @ts-ignore
	if (extraSettings) args[2] = { ...args[2], ...extraSettings }

	return args as any
}
