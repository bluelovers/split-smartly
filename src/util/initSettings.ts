import { IIncludeSeparatorMode, ISplitSettings } from '../types';
import { isEmpty } from '../util';

export function initSettings<M extends IIncludeSeparatorMode>(settings: ISplitSettings<M>)
{
	if (Array.isArray(settings.mentions) || typeof settings.mentions === 'string')
	{
		const mentionsMap =
			[settings.mentions].flat().filter(Boolean).reduce((map, keyword: string) =>
			{
				const key = settings.ignoreCase ? keyword.toUpperCase() : keyword
				map[key] = keyword
				return map
			}, {} as Record<string, string>)

		settings.mentions = !isEmpty(mentionsMap) && mentionsMap
	}

	return settings
		.createBracketsMap()
		.createBracketsSearch()
		.createSeparatorsSearch()
}
