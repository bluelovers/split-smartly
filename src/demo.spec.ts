import { IIncludeSeparatorMode, IParametersSplitSmartly } from './types';
import splitSmartly from './index';
import { inspect } from 'util';
// @ts-ignore
import { fixtureDemo } from '../test/fixture/demo';

describe(`demos`, () =>
{

	fixtureDemo
		.forEach((argv) => {

			test(inspect(argv), () =>
			{
				let actual = splitSmartly(...argv as any);

				expect(actual).toMatchSnapshot();
			});

		})
	;

})
