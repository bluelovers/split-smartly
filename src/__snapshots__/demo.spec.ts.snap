// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`demos [
  '(one / two) / "three / four" / <<five / six>>',
  '/',
  { brackets: '(), << >>' }
] 1`] = `
[
  "(one / two)",
  ""three / four"",
  "<<five / six>>",
]
`;

exports[`demos [
  '(one / two) / "three / four" / <<five / six>>',
  '/',
  { brackets: [ [Array], [Array] ] }
] 1`] = `
[
  "(one / two)",
  ""three / four"",
  "<<five / six>>",
]
`;

exports[`demos [
  '(one / two) / "three / four" / <<five / six>>',
  '/',
  { brackets: { '(': ')', '<<': '>>' } }
] 1`] = `
[
  "(one / two)",
  ""three / four"",
  "<<five / six>>",
]
`;

exports[`demos [
  'Peter loves Mary and Mary loves Johnny and Jonny loves Steve',
  'AND',
  { mentions: [ 'STEVE', 'PETER' ] }
] 1`] = `
[
  {
    "mentions": [
      "PETER",
    ],
    "text": "Peter loves Mary",
  },
  {
    "text": "Mary loves Johnny",
  },
  {
    "mentions": [
      "STEVE",
    ],
    "text": "Jonny loves Steve",
  },
]
`;

exports[`demos [
  'SELECT best FROM life',
  [ 'SELECT ', 'FROM ' ],
  { includeSeparatorMode: 'ONLY' }
] 1`] = `
[
  "SELECT ",
  "FROM ",
]
`;

exports[`demos [
  'SELECT best FROM life',
  [ 'SELECT ', 'FROM ' ],
  { includeSeparatorMode: 'RIGHT' }
] 1`] = `
[
  [
    "SELECT ",
    "best",
  ],
  [
    "FROM ",
    "life",
  ],
]
`;

exports[`demos [
  'SELECT best FROM life',
  [ 'SELECT ', 'FROM ' ],
  { includeSeparatorMode: 'SEPARATELY' }
] 1`] = `
[
  "",
  "SELECT ",
  "best",
  "FROM ",
  "life",
]
`;

exports[`demos [
  'select best FROM life',
  [ 'SELECT ', 'FROM ' ],
  { includeSeparatorMode: 'RIGHT' }
] 1`] = `
[
  [
    "SELECT ",
    "best",
  ],
  [
    "FROM ",
    "life",
  ],
]
`;

exports[`demos [
  'select best FROM life',
  [ 'SELECT ', 'FROM ' ],
  { includeSeparatorMode: 'RIGHT', ignoreCase: false }
] 1`] = `
[
  [
    undefined,
    "select best",
  ],
  [
    "FROM ",
    "life",
  ],
]
`;

exports[`demos [
  {
    brackets: [],
    mentions: [],
    ignoreInsideQuotes: true,
    includeSeparatorMode: 'NONE',
    ignoreCase: true,
    trimResult: true,
    trimSeparators: false,
    defaultBrackets: [ [Array], [Array], [Array] ]
  }
] 1`] = `[Function]`;

exports[`demos [ 'One | Two | Three | Four', '|', { indexes: [ 1, 3 ] } ] 1`] = `
[
  "Two",
  "Four",
]
`;

exports[`demos [ 'One | Two | Three | Four', '|', { indexes: 2 } ] 1`] = `"Three"`;

exports[`demos [ 'SELECT best FROM life', [ 'SELECT ', 'FROM ' ] ] 1`] = `
[
  "",
  "best",
  "life",
]
`;

exports[`demos [ 'one , two , "three , four" , five , six' ] 1`] = `
[
  "one",
  "two",
  ""three , four"",
  "five",
  "six",
]
`;

exports[`demos [ 'one _ two _ "three _ four" _ five _ six', '_' ] 1`] = `
[
  "one",
  "two",
  ""three _ four"",
  "five",
  "six",
]
`;

exports[`demos [ 'one _ two _ "three _ four" _ five _ six', '_', { brackets: true } ] 1`] = `
[
  "one",
  "two",
  ""three _ four"",
  "five",
  "six",
]
`;

exports[`demos [ 'p1: point first p2: point second', /p\\d:/gi ] 1`] = `
[
  "",
  "point first",
  "point second",
]
`;

exports[`demos [ [ 'SELECT ', 'FROM ' ] ] 1`] = `[Function]`;

exports[`demos [ { includeSeparatorMode: 'RIGHT', ignoreCase: false } ] 1`] = `[Function]`;

exports[`demos [ {} ] 1`] = `[Function]`;
